File: Arrays and Hashing/706-Design-HashMap/706.js
var ListNode = function(key, value, next) {
    this.key = key;
    this.value = value;
    this.next = next;
}

var MyHashMap = function() {
    this.size = 19997;
    this.mult = 12582917;
    this.data = new Array(this.size);
};

MyHashMap.prototype.hash = function(key){
    return key * this.mult % this.size;
}

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
MyHashMap.prototype.put = function(key, value) {
    this.remove(key);
    let h = this.hash(key);
    let node = new ListNode(key, value, this.data[h]);
    this.data[h] = node;
};

/** 
 * @param {number} key
 * @return {number}
 */
MyHashMap.prototype.get = function(key) {
    let h = this.hash(key), node = this.data[h];
    while (node){
        if (node.key === key){
            return node.value;
        }
        node = node.next;
    }
    return -1;
};

/** 
 * @param {number} key
 * @return {void}
 */
MyHashMap.prototype.remove = function(key) {
    let h = this.hash(key), node = this.data[h]
        if (!node) return
        if (node.key === key) this.data[h] = node.next
        else while (node.next){
            if (node.next.key === key) {
                node.next = node.next.next
                return
            }
            node = node.next
        }
};

/** 
 * Your MyHashMap object will be instantiated and called as such:
 * var obj = new MyHashMap()
 * obj.put(key,value)
 * var param_2 = obj.get(key)
 * obj.remove(key)
 */

File: Arrays and Hashing/706-Design-HashMap/706.java
class ListNode {
    int key, value;
    ListNode next;
    public ListNode(int key, int value, ListNode next) {
        this.key = key;
        this.value = value;
        this.next = next;
    }
}
class MyHashMap {
    static final int size = 19997;
    static final int mult = 12582917;
    ListNode[] data;
    public MyHashMap() {
        this.data = new ListNode[size];
    }
    private int hash(int key) {
        return (int)((long)key * mult % size);
    }
    public void put(int key, int value) {
        remove(key);
        int h = hash(key);
        ListNode node = new ListNode(key, value, data[h]);
        data[h] = node;
    }
    public int get(int key) {
        int h = hash(key);
        ListNode node = data[h];
        while (node != null){
            if (node.key == key){
                return node.value;
            }
            node = node.next;
        }
        return -1;
    }
    public void remove(int key) {
        int h = hash(key);
        ListNode node = data[h];
        if (node == null) return;
        if (node.key == key) data[h] = node.next;
        else while(node != null){
            if (node.next.key == key) {
                node.next = node.next.next;
                return;
            }
            node = node.next;
        }
    }
}
/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.put(key,value);
 * int param_2 = obj.get(key);
 * obj.remove(key);
 */